pkg/storage/registry/spaces/spaces.go

```diff
@@ -0,0 +1,457 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package spaces
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"text/template"
+
+	"github.com/Masterminds/sprig"
+	userpb "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	rpc "github.com/cs3org/go-cs3apis/cs3/rpc/v1beta1"
+	providerpb "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	registrypb "github.com/cs3org/go-cs3apis/cs3/storage/registry/v1beta1"
+	typesv1beta1 "github.com/cs3org/go-cs3apis/cs3/types/v1beta1"
+	"github.com/cs3org/reva/pkg/appctx"
+	ctxpkg "github.com/cs3org/reva/pkg/ctx"
+	"github.com/cs3org/reva/pkg/errtypes"
+	"github.com/cs3org/reva/pkg/logger"
+	"github.com/cs3org/reva/pkg/rgrpc/status"
+	"github.com/cs3org/reva/pkg/rgrpc/todo/pool"
+	"github.com/cs3org/reva/pkg/sharedconf"
+	"github.com/cs3org/reva/pkg/storage"
+	pkgregistry "github.com/cs3org/reva/pkg/storage/registry/registry"
+	"github.com/mitchellh/mapstructure"
+	"google.golang.org/grpc"
+)
+
+//go:generate mockery -name StorageProviderClient
+
+func init() {
+	pkgregistry.Register("spaces", NewDefault)
+}
+
+type provider struct {
+	Mapping           string            `mapstructure:"mapping"`
+	MountPath         string            `mapstructure:"mount_path"`
+	Aliases           map[string]string `mapstructure:"aliases"`
+	AllowedUserAgents []string          `mapstructure:"allowed_user_agents"`
+	PathTemplate      string            `mapstructure:"path_template"`
+	template          *template.Template
+	// filters
+	SpaceType      string `mapstructure:"space_type"`
+	SpaceOwnerSelf bool   `mapstructure:"space_owner_self"`
+	SpaceID        string `mapstructure:"space_id"`
+}
+
+type templateData struct {
+	CurrentUser *userpb.User
+	Space       *providerpb.StorageSpace
+}
+
+// StorageProviderClient is the interface the spaces registry uses to interact with storage providers
+type StorageProviderClient interface {
+	ListStorageSpaces(ctx context.Context, in *providerpb.ListStorageSpacesRequest, opts ...grpc.CallOption) (*providerpb.ListStorageSpacesResponse, error)
+}
+
+// WithSpace generates a layout based on space data.
+func (p *provider) ProviderPath(u *userpb.User, s *providerpb.StorageSpace) (string, error) {
+	b := bytes.Buffer{}
+	if err := p.template.Execute(&b, templateData{CurrentUser: u, Space: s}); err != nil {
+		return "", err
+	}
+	return b.String(), nil
+}
+
+type config struct {
+	Providers    map[string]*provider `mapstructure:"providers"`
```

Comment:
this abstraction leaks. I would either work with an interface that refers to `provider` or make `provider` public, but this reads a bit confusing, since you're exporting `Providers` which is composed of a map to a private type. Worth keeping it in mind for the next iteration.

Reply:
I'm guessing `Providers` is public because we need to unmarshal things onto it

Reply:
but `provider` struct is private, yet all its fields are exported 🤔 this is a red flag

Reply:
Made everything public in https://github.com/cs3org/reva/pull/2383